Update IPC pages (#1949)
filecoin-project/filecoin-docs: content/en/basics/interplanetary-consensus/mir-and-trantor/index.md

---
title: "Mir and Trantor"
description: "Alongside the scaling framework, the IPC project is developing high-performance consensus protocols that are suitable for subnet use."
lead: "Alongside the scaling framework, the IPC project is developing high-performance consensus protocols that are suitable for subnet use."
draft: false
images: []
type: docs
menu:
  basics:
    parent: "basics-interplanetary-consensus"
    identifier: "ipc-mir-and-trantor"
weight: 120
toc: true
---

Beyond their function as a scaling tool, IPC subnets will be deployed to meet the requirements of applications that cannot be run on the Filecoin rootnet. The combination of Mir and Trantor provides a high-performance consensus mechanism with faster finality and higher throughput. Moreover, Trantor can be further configured for each subnet.

## Mir framework

> Mir is a framework for implementing, debugging, and analyzing distributed protocols. It takes the form of a library that provides abstractions representing different components of a distributed system and an engine orchestrating their interaction.

Reviewer: ```suggestion
[Mir](https://github.com/filecoin-project/mir) is a framework for implementing, debugging, and analyzing distributed protocols. It is organized as a library that provides abstractions representing different components of a distributed system, as well as an engine to orchestrate their interaction.
```

> Mir aims to be general enough to enable implementing distributed protocols in a way that is agnostic to network transport, storage, and cryptographic primitive implementation. All these (and other) usual components of a distributed protocol implementation are encapsulated in abstractions with well-defined interfaces. While Mir provides some implementations of those abstractions to be used directly "out of the box", the consumer is free to provide their own implementations.
> 
> The first use of Mir is as a scalable and efficient consensus layer in Filecoin subnets. Learn more on [GitHub](https://github.com/filecoin-project/mir).
> 

Reviewer: ```suggestion
The goal of Mir is to enable implementation of distributed protocols in a way that is agnostic to network transport, storage, and cryptographic primitive implementation. All of these components of a distributed protocol implementation, as well as others, are encapsulated in abstractions with well-defined interfaces. While Mir provides some implementations of abstractions  for direct use "out of the box", the user is free to define their own implementations.

Mir is currently being used as a scalable and efficient consensus layer in Filecoin subnets. Learn more at [GitHub](https://github.com/filecoin-project/mir).
```

Author: I like what you did but would further change the final sentence to

> Mir is used as a scalable and efficient consensus layer in IPC subnets.

Author: ```suggestion
The goal of Mir is to enable implementation of distributed protocols in a way that is agnostic to network transport, storage, and cryptographic primitive implementation. All of these components of a distributed protocol implementation, as well as others, are encapsulated in abstractions with well-defined interfaces. While Mir provides some implementations of abstractions  for direct use "out of the box", the user is free to define their own implementations.

Mir is used as a scalable and efficient consensus layer in IPC subnets.. Learn more at [GitHub](https://github.com/filecoin-project/mir).
```
(your suggestion with tweaks)

I like what you did but would further change that last paragraph. Reasoning: currently suggests transient, which isn't the case, and "Filecoin subnets" is poorly defined (mea culpa).

## Trantor consensus protocol

> Trantor is a modern multi-leader Byzantine fault-tolerant (BFT) protocol, inspired by recently proposed high-throughput BFT consensus protocols, namely ISS and Narwahl. It iterates through instances of PBFT with immediate finality, each block containing an ordered list of decided transactions and a certificate for verification, with every ∆-th block containing a checkpoint of the state.

Reviewer: ```suggestion
[Trantor](https://hackmd.io/P59lk4hnSBKN5ki5OblSFg?view) is a modern, multi-leader, Byzantine fault-tolerant (BFT) protocol,. It was inspired by recently proposed, high-throughput, BFT consensus protocols, namely ISS and Narwahl. Trantor iterates through instances of PBFT with immediate finality, each block containing an ordered list of decided transactions and a certificate for verification, with every ∆-th block containing a checkpoint of the state.
```

Reviewer: As a reader, it would not be clear to me what `∆-th` and `PBFT ` is - need definitions

Author: ```suggestion
Trantor is a modern, multi-leader, Byzantine fault-tolerant (BFT) protocol. It was inspired by recently proposed, high-throughput, BFT consensus protocols, namely ISS and Narwahl. Trantor iterates through instances of PBFT with immediate finality, each block containing an ordered list of decided transactions and a certificate for verification, with every nth block containing a checkpoint of the state.
```

I intentionally dropped the hackmd as that was an old working document. Immediately below, we have a link to the github readme, which is the appropriate reference. I don't think it's useful to link twice, but fine if you prefer it that way. There was a typo (`,.`) in the suggestion. `∆-th` changed to `nth`. I'm not sure what to do about PBFT: it's a foundational protocol in distributed systems, and not something that we can define (the P stands for practical, but that won't help anyone who doesn't already know of it).

> Learn more on [GitHub](https://github.com/filecoin-project/mir/tree/main/pkg/systems/trantor). 

Author: ```suggestion
Learn more on [GitHub](https://github.com/filecoin-project/mir/tree/main/pkg/trantor). 
```

> To keep up with the development of Mir and Trantor, consider joining the `#mir-dev` channel in the  [Filecoin Slack workspace](https://filecoin.io/slack/). 

Reviewer: ```suggestion
To stay updated on Mir and Trantor development, join the `#mir-dev` channel in the  [Filecoin Slack workspace](https://filecoin.io/slack/). 
```
